engine/access/rpc/backend/transaction_subscription_metadata.go

```diff
@@ -0,0 +1,254 @@
+		txReferenceBlockID:   txReferenceBlockID,
+	}
+}
+
+// Refresh updates the transaction subscription metadata to reflect the latest state.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//
+// Expected errors during normal operation:
+//   - [ErrBlockNotReady] if the block at the given height is not found.
+//   - codes.Internal if impossible to get transaction result due to event payload conversion failed
+//
+// All other errors are considered as state corruption (fatal) or internal errors in the refreshing transaction result
+// or when refreshing transaction status.
+func (tm *transactionSubscriptionMetadata) Refresh(ctx context.Context) error {
+	if err := tm.refreshCollection(); err != nil {
+		return err
+	}
+	if err := tm.refreshBlock(); err != nil {
+		return err
+	}
+	if err := tm.refreshTransactionResult(ctx); err != nil {
+		return err
+	}
+	if err := tm.refreshStatus(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// refreshTransactionReferenceBlockID sets the reference block ID for the transaction.
+//
+// If the reference block ID is unset, it attempts to retrieve it from storage.
+//
+// Parameters:
+//   - txReferenceBlockID: The reference block ID of the transaction.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) refreshTransactionReferenceBlockID() error {
+	// Get referenceBlockID if it is not set
+	if tm.txReferenceBlockID != flow.ZeroID {
+		return nil
+	}
+
+	tx, err := tm.transactions.ByID(tm.txResult.TransactionID)
+	if err != nil {
+		return fmt.Errorf("failed to lookup transaction by transaction ID: %w", err)
+	}
+	tm.txReferenceBlockID = tx.ReferenceBlockID
+	return nil
+}
+
+// refreshStatus updates the transaction's status based on its execution result.
+//
+// Parameters:
+//   - ctx: Context for managing the operation lifecycle.
+//
+// No errors expected during normal operations.
+func (tm *transactionSubscriptionMetadata) refreshStatus(ctx context.Context) error {
+	var err error
+
+	if tm.blockWithTx == nil {
+		if err = tm.refreshTransactionReferenceBlockID(); err != nil {
+			// transaction was not sent from this node, and it has not been indexed yet.
+			if errors.Is(err, storage.ErrNotFound) && tm.txReferenceBlockID == flow.ZeroID {
+				tm.txResult.Status = flow.TransactionStatusUnknown
+				return nil
+			}
+			return err
+		}
+
+		tm.txResult.Status, err = tm.backendTransactions.DeriveUnknownTransactionStatus(tm.txReferenceBlockID)
+		if err != nil {
+			if !errors.Is(err, state.ErrUnknownSnapshotReference) {
+				irrecoverable.Throw(ctx, err)
+			}
+			return rpc.ConvertStorageError(err)
+		}
+		return nil
+	}
+
+	// When a block with the transaction is available, it is possible to receive a new transaction status while
+	// searching for the transaction result. Otherwise, it remains unchanged. So, if the old and new transaction
+	// statuses are the same, the current transaction status should be retrieved.
```

Comment:
I'm confused about what this comment means. Here's my understanding. feel free to use it or reword.
```suggestion
	// When the transaction is included in an executed block, the `txResult` may be updated during `Refresh`
	// Recheck the status to ensure it's accurate.
```